modules/navigation/editor/navigation_mesh_editor_plugin.cpp

```diff
@@ -60,67 +68,101 @@ void NavigationMeshEditor::_notification(int p_what) {
-					err_dialog->set_text(TTR("Cannot generate navigation mesh because it belongs to a resource which was imported."));
-					err_dialog->popup_centered();
-					return;
-				}
-			}
-		}
-	} else {
-		if (FileAccess::exists(path + ".import")) {
-			err_dialog->set_text(TTR("Cannot generate navigation mesh because the resource was imported from another type."));
+	HashSet<Ref<NavigationMesh>> navmeshes;
+	LocalVector<NavigationRegion3D *> regions_with_navmesh_to_bake;
+
+	for (NavigationRegion3D *region : selected_regions) {
+		ERR_CONTINUE(region == nullptr);
+		Ref<NavigationMesh> navmesh = region->get_navigation_mesh();
+		if (navmesh.is_null()) {
+			err_dialog->set_text(TTR("A NavigationMesh resource must be set or created for this node to work."));
 			err_dialog->popup_centered();
 			return;
 		}
+
+		if (navmeshes.has(navmesh)) {
+			// No point (re)baking the same resource in case of multi select.
+			// Trying to bake the same navmesh twice would trigger an error.
+			continue;
+		}
+		navmeshes.insert(navmesh);
+
+		String path = navmesh->get_path();
+		if (!path.is_resource_file()) {
+			int srpos = path.find("::");
+			if (srpos != -1) {
+				String base = path.substr(0, srpos);
+				if (ResourceLoader::get_resource_type(base) == "PackedScene") {
+					if (!get_tree()->get_edited_scene_root() || get_tree()->get_edited_scene_root()->get_scene_file_path() != base) {
+						err_dialog->set_text(TTR("Cannot generate navigation mesh because it does not belong to the edited scene. Make it unique first."));
+						err_dialog->popup_centered();
+						return;
+					}
+				} else {
+					if (FileAccess::exists(base + ".import")) {
+						err_dialog->set_text(TTR("Cannot generate navigation mesh because it belongs to a resource which was imported."));
+						err_dialog->popup_centered();
+						return;
+					}
+				}
+			}
+		} else {
+			if (FileAccess::exists(path + ".import")) {
+				err_dialog->set_text(TTR("Cannot generate navigation mesh because the resource was imported from another type."));
+				err_dialog->popup_centered();
+				return;
+			}
+		}
+		regions_with_navmesh_to_bake.push_back(region);
 	}
 
-	node->bake_navigation_mesh(true);
+	// Only use a bake thread when user has a single region selected.
+	// There is a high risk crashing the editor from running out of memory
+	// should user select and bake many heavy navmeshes at the same time.
+	// Can't precalculate memory cost without knowing the source geometry.
+	bool use_bake_thread = regions_with_navmesh_to_bake.size() == 1;
+
+	for (NavigationRegion3D *region : regions_with_navmesh_to_bake) {
+		region->bake_navigation_mesh(use_bake_thread);
+	}
 
-	node->update_gizmos();
+	for (NavigationRegion3D *region : selected_regions) {
+		// Set navmesh again to trigger update of the debug visuals also
+		// for all regions that have not baked because they share navmesh.
+		region->set_navigation_mesh(region->get_navigation_mesh());
+		region->update_gizmos();
+	}
```

Comment:
`bake_navigation_mesh` calls `_bake_finished`, which calls `set_navigation_mesh`, which calls `update_gizmos`.  
Just from looking at the code, this for-loop should not be necessary…

```suggestion
```

Reply:
The bake_navigation_mesh() is only called for the first region with a unique navmesh, not for any other region that shares the same navmesh as it would trigger "this navmesh is already baking" errors. So this is done so that all the regions update their debug because they would not receive a callback.